src/internal/curry.mjs

```diff
@@ -17,40 +17,47 @@ function unshiftUndefineds (args, by) {
   }
 }
 
-class BaseIterable {
-  constructor (fn, variadic, args, iterableArgs) {
-    this._fn = fn
-    this._args = args
-    this._variadic = variadic
-    this._iterableArgs = iterableArgs
-    this._staticIterator = null
-  }
+function BaseIterable (fn, variadic, args, iterableArgs) {
+  this._fn = fn
+  this._args = args
+  this._variadic = variadic
+  this._iterableArgs = iterableArgs
+  this._staticIterator = null
+}
```

Comment:
I think you can make it even shorter. (this is merely converting ES6 classes to ES5 constructor functions, and you still use `this`)

Anyway, if you still don't know what I meant, I'll write it tomorrow.

Reply:
If you intend to replace an instance with some pattern involving closures I think you will be sacrificing runtime perf at the expense of a very small amount of bundle size. For the most core part of the libraries it is important that we create functions which can get hot and be compiled to native code. This means that the dynamic things we work with should be objects, not functions, and should have a stable shape.

Reply:
It sounds like a premature optimization to me, and it requires you to predict (or assume) the internal working of all JavaScript engines, which are unreliable and always changing. And you don't even have a proper benchmark.

The only thing that I am certain of is bundle size and readability, and I find the closure approach satisfies both.

That is not to say that performance is unnecessary. It is. But without benchmark, we only take a shot in the dark.

Reply:
May I suggest that instead of throwing your arms up and declaring that it is impossible to understand how javascript engines work and that they having nothing in common, you do some studying on the topic? I recommend you start [here](https://mathiasbynens.be/notes/shapes-ics) and continue with [this](https://mathiasbynens.be/notes/prototypes)

Reply:
Your provided links do not invalidate my arguments which is:
  1. Avoid premature optimization
  2. Engines are always changing

Beside, the links you provided all point to a single V8 engine. There're also SpiderMonkey, JavaScriptCore, etc. That is one of the reasons why benchmark is needed when performance is concern.

And what not to say reducing bundle size does not gives you performance benefit? Download time, parsing time.

Reply:
The links note the similarities between many engines, not just v8. They *all* have optimizing compilers. Out of curiosity, do you know C or C++? Have you dealt with structs and pointer arithmetic, for example? There is much that is not always changing about the way that computers work. It is by writing code that mirrors the low level way a processor actually does work that you avoid the uncertainty of how various JS engines treat the more dynamic patterns that javascript allows.

Reply:
But what is certain is that if you create a function many times, even if it seems like one function in your mind, it is not to the compiler. The compiler cannot optimize repeated calls to the same function unless the repeated calls are actually to the same function. This is basic stuff, not compiler internal rocket science. There *may* be some compiler internal rocket science that can reverse engineer that multiple methods really should be the same, but that would be the kind of thing that we would be always changing and we would definitely not want to rely on.

Reply:
I'm going to reverse your argument on you though. You're talking about making sacrifices to correctness to gain 1KB of raw bundle size, and less when compression is in use! That's premature optimization if I've ever heard of it.

Reply:
> "Premature optimization" is a phrase used to describe a situation where a programmer lets performance considerations affect the design of a piece of code. This can result in a design that is not as clean as it could have been or code that is incorrect, because the code is complicated by the optimization and the programmer is distracted by optimizing.

(From https://en.wikipedia.org/wiki/Program_optimization#When_to_optimize)

Can we agree on this? Correctness first, optimization second? Not to say that we should accept solutions which are the wrong complexity class or code which is twice as slow as it should be, but correctness should always come first, and should be *followed* by optimization if necessary.